### ISSUE TITLE
Run hourly between specific time frame everyday

### ISSUE THREAD

**Issue by yoonwaiyan** [rossta/montrose#119]
Hi, I've tried dozens of combinations to make montrose gem able to generate the list of recurring events that runs hourly between a specific time frame everyday. I'm expecting `Montrose.every(30.minutes, hour: 9..16)` to return events starting from 9am but it shows the current time as the latest event instead and intervals are counted based on current time: 

```
=> [2019-07-31 14:29:46 +0800,
 2019-07-31 14:59:46 +0800,
 2019-07-31 15:29:46 +0800,
 2019-07-31 15:59:46 +0800,
 2019-07-31 16:29:46 +0800,
 2019-07-31 16:59:46 +0800,
 2019-08-01 09:29:46 +0800]
```

While `Montrose.every(30.minutes, starts: "9:00 AM", hour: 9..16)` works, I want to have the set of events that user could set to start and end at anytime instead of a rounded number. In this case, maybe user wants to run the recurring events from 9AM to 4PM everyday, so the final event of a day should happen before 4PM. This is a fraction of the events I captured:

```
[2019-07-31 14:00:00 +0800,
 2019-07-31 14:30:00 +0800,
 2019-07-31 15:00:00 +0800,
 2019-07-31 15:30:00 +0800,
 2019-07-31 16:00:00 +0800,
 2019-07-31 16:30:00 +0800,
 2019-08-01 09:00:00 +0800,
 2019-08-01 09:30:00 +0800,
 2019-08-01 10:00:00 +0800,]
```

The goal is to let user select a time range they would like to run their hourly update, e.g. run a process everyday between 9:15am and 5:15pm every 2 hours. What's the best way to input this to Montrose gem?

**Comment by rossta** [comment]
I think the `between` method might be what you're looking for:

```ruby
$ bin/console
[1] pry(main)> Montrose.every(30.minutes).between(Time.parse("9:15AM")..Time.parse("5:15PM")).to_a
=> [2019-07-31 09:15:00 -0400,
 2019-07-31 09:45:00 -0400,
 2019-07-31 10:15:00 -0400,
 2019-07-31 10:45:00 -0400,
 2019-07-31 11:15:00 -0400,
 2019-07-31 11:45:00 -0400,
 2019-07-31 12:15:00 -0400,
 2019-07-31 12:45:00 -0400,
 2019-07-31 13:15:00 -0400,
 2019-07-31 13:45:00 -0400,
 2019-07-31 14:15:00 -0400,
 2019-07-31 14:45:00 -0400,
 2019-07-31 15:15:00 -0400,
 2019-07-31 15:45:00 -0400,
 2019-07-31 16:15:00 -0400,
 2019-07-31 16:45:00 -0400,
 2019-07-31 17:15:00 -0400]
```

Let me know if that helps.

**Comment by rossta** [comment]
Correction, the `between` method won't repeat the recurrence every day. On re-reading your post, you could set `starts` time to be `9:15am` and using the exclusive range `...` to get the times *before* 4pm:

```ruby
enum = Montrose.every(30.minutes, starts: "9:15AM", hour: 9...16)
=> #<Montrose::Recurrence:3ffee8d2c52c {:every=>:minute, :starts=>2019-07-31 09:15:00 -0400, :hour=>[9, 10, 11, 12, 13, 14, 15], :interval=>30}>
[33] pry(main)> enum.take(15)
=> [2019-07-31 09:15:00 -0400,
 2019-07-31 09:45:00 -0400,
 2019-07-31 10:15:00 -0400,
 2019-07-31 10:45:00 -0400,
 2019-07-31 11:15:00 -0400,
 2019-07-31 11:45:00 -0400,
 2019-07-31 12:15:00 -0400,
 2019-07-31 12:45:00 -0400,
 2019-07-31 13:15:00 -0400,
 2019-07-31 13:45:00 -0400,
 2019-07-31 14:15:00 -0400,
 2019-07-31 14:45:00 -0400,
 2019-07-31 15:15:00 -0400,
 2019-07-31 15:45:00 -0400,
 2019-08-01 09:15:00 -0400]
```
It doesn't look like this will work if you want the last time to be exactly at 4pm. 

For a more precise handling of minutely and hourly recurrences during specific times of day, we could add a `During` rule which would work similarly to `HourOfDay`.

Maybe it could work something like the following:

```ruby
Montrose.every(30.minutes).during(Time.parse('9:00am')..Time.parse('4:00pm')).starts('9:15am')
```
and even accept multiple ranges:
```ruby
range_1 = Time.parse('9:00am')..Time.parse('11:00am')
range_2 = Time.parse('1:00pm')..Time.parse('3:00pm')
range_3 = Time.parse('5:00pm')..Time.parse('7:00pm')
Montrose.every(30.minutes).during(range_1, range_2, range_3).starts('9:15am')
```
What do you think?

**Comment by yoonwaiyan** [comment]
Thanks for the pointers. I observed that time ranges are normally handled in such a way that it ends before :00 (so it's `3:59PM` in this case for events that should ends before 4PM), so it's ok as long as montrose return events that happens between and inclusive of the time defined by users, which is why I was puzzled when I see `2019-07-31 16:30:00 +0800` in the result I queried.  Thanks for pointing out the `...`, never thought of that but I would like to call this directly from users' input so `Time.parse` is definitely a good idea, but is `during` a new feature? Never seen this in any part of the codebase though.

**Comment by yoonwaiyan** [comment]
I tried to query based on your result using `between` but it seemed to only return events for a particular day instead of infinite number of events:

```ruby
[6] pry(main)> Montrose.every(30.minutes).between(Time.parse('9:00am')..Time.parse('3:59pm')).starts('9:15am').events.take(31)
=> [2019-08-01 09:15:00 +0800,
 2019-08-01 09:45:00 +0800,
 2019-08-01 10:15:00 +0800,
 2019-08-01 10:45:00 +0800,
 2019-08-01 11:15:00 +0800,
 2019-08-01 11:45:00 +0800,
 2019-08-01 12:15:00 +0800,
 2019-08-01 12:45:00 +0800,
 2019-08-01 13:15:00 +0800,
 2019-08-01 13:45:00 +0800,
 2019-08-01 14:15:00 +0800,
 2019-08-01 14:45:00 +0800,
 2019-08-01 15:15:00 +0800,
 2019-08-01 15:45:00 +0800]
```

**Comment by rossta** [comment]
> between but it seemed to only return events for a particular day instead of infinite number of events

Correct, I misread your use case for my first response.

> is during a new feature?

`during` doesn't exist, it's a proposed solution to handle use cases like the one you've described. I'd be interested to see if it would work for your needs.

**Comment by yoonwaiyan** [comment]
Yes that would be great, thanks!

**Comment by rossta** [comment]
Check out #121 and let me know if you have any feedback

**Comment by yoonwaiyan** [comment]
Looks good 🙂 it would be great to have this documented in README too.

**Comment by rossta** [comment]
Thanks for your input. I've merged #121. Try it out on master if you're able to and let me know if you have any issues or questions.

**Comment by rossta** [comment]
Now merged into master.

**Comment by yoonwaiyan** [comment]
@rossta tested `during` and it's good for my needs for now. The only thing left is a gem release so I can update my gem dependency properly.
